File: spark-scala-maven/src/main/scala/adj/spark/core/advance/Top3.scala
package adj.spark.core.advance

import org.apache.spark.SparkConf
import org.apache.spark.SparkContext
import scala.collection.mutable.ArrayBuffer

object Top3 {
  def main(args: Array[String]): Unit = {
    val conf = new SparkConf()
               .setAppName("Top3")
               .setMaster("local")
    val sc = new SparkContext(conf)
    
    val lineRDD = sc.textFile("hdfs://master:9000/user/spark/data/sort.txt", 3)
    val numberRDD = lineRDD.flatMap(line => {
      val wordList = ArrayBuffer[Int]()
      for(word <- line.split(" ")) {
        wordList += word.toInt
      }
      wordList
    })
    val numberPairRDD = numberRDD.map(number => (number, 1))
    val sortedNumberRDD = numberPairRDD.sortByKey(false, 3)
    val tackNumberList  = sortedNumberRDD.take(3)
    
    for(tackNumber <- tackNumberList) {
      println("number = " + tackNumber._1)
    }
  }
}

File: spark-java/src/main/java/adj/spark/core/advance/Top3.java
package adj.spark.core.advance;

import java.util.ArrayList;
import java.util.List;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.FlatMapFunction;
import org.apache.spark.api.java.function.Function;
import org.apache.spark.api.java.function.PairFunction;

import scala.Tuple2;

public class Top3 {
	public static void main(String[] args) {
		SparkConf conf = new SparkConf()
				         .setAppName("Top3")
				         .setMaster("local");
		@SuppressWarnings("resource")
		JavaSparkContext jsc = new JavaSparkContext(conf);
		
		JavaRDD<String> lineRDD = jsc.textFile("hdfs://master:9000/user/spark/data/sort.txt", 3);
		
		JavaRDD<Integer> numberRDD = lineRDD.flatMap(new FlatMapFunction<String, Integer>() {
			private static final long serialVersionUID = 1L;

			@Override
			public Iterable<Integer> call(String line) throws Exception {
				List<Integer> numberList = new ArrayList<>();
				for(String word: line.split(" ")) {
					numberList.add(Integer.valueOf(word));
				}
				return numberList;
			}
		});
		JavaPairRDD<Integer, Integer> numberPairRDD = numberRDD.mapToPair(new PairFunction<Integer, Integer, Integer>() {
			private static final long serialVersionUID = 1L;

			@Override
			public Tuple2<Integer, Integer> call(Integer number) throws Exception {
				return new Tuple2<Integer, Integer>(number, 1);
			}
		});
		JavaPairRDD<Integer, Integer> sortedNumberRDD = numberPairRDD.sortByKey(false);
		JavaRDD<Integer> numberMapRDD = sortedNumberRDD.map(new Function<Tuple2<Integer,Integer>, Integer>() {
			private static final long serialVersionUID = 1L;

			@Override
			public Integer call(Tuple2<Integer, Integer> sortedNumber) throws Exception {
				return sortedNumber._1;
			}
		});
		List<Integer> tackNumberList = numberMapRDD.take(3);
		for(Integer number: tackNumberList) {
			System.out.println("number = " + number);
		}
		jsc.close();
	}
}
